internal/btf/co_re.go

```diff
@@ -0,0 +1,737 @@
+package btf
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/ebpf/internal"
+)
+
+var (
+	// ErrInvalidCORESpec indicates an invalid CO-RE relocation spec was parsed
+	ErrInvalidCORESpec        = errors.New("invalid CO-RE spec")
+	errStructureNeedsCleaning = errors.New("structure needs cleaning")
+	errUnsupportedRelocation  = errors.New("unsupported CO-RE relocation")
+)
+
+// LoadCORERelocations parses the relocations from the BTF extended ELF section and calculates how to apply
+// the relocation to the target, described by the BTF provided. If no BTF is provided, the current kernel BTF
+// will attempted to be loaded. Returns a mapping between instruction offset and the relocation, with result.
+func (s *Spec) LoadCORERelocations(name string, targetBtf *Spec, bo binary.ByteOrder) (map[uint64]CORERelocation, error) {
+	relos, ok := s.reloInfos[name]
+	if !ok || len(relos.records) == 0 {
+		return nil, nil
+	}
+
+	rs := map[uint64]CORERelocation{}
+	cache := map[Type][]Type{}
+	var err error
+
+	if targetBtf == nil {
+		targetBtf, err = LoadKernelSpec()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get target btf: %w", err)
+		}
+	}
+
+	for _, r := range relos.records {
+		cr := coreReloRecord{}
+		if err := binary.Read(bytes.NewReader(r.Opaque), bo, &cr); err != nil {
+			return nil, fmt.Errorf("unable to read CO-RE relocation record: %w", err)
+		}
+
+		relo := CORERelocation{
+			Type: s.indexedTypes[cr.TypeId],
+			Kind: cr.ReloKind,
+		}
+		relo.Accessor, err = s.strings.Lookup(cr.AccessStrOff)
+		if err != nil {
+			return nil, fmt.Errorf("invalid CO-RE relocation accessor string: %w", err)
+		}
+		relo.Result, err = calculateRelocation(&relo, targetBtf, cache)
+		if err != nil {
+			return nil, fmt.Errorf("failed to calculate CO-RE relocation: %w", err)
+		}
+
+		rs[r.InsnOff] = relo
+	}
+	return rs, nil
+}
+
+func calculateRelocation(relo *CORERelocation, targetBtf *Spec, candCache map[Type][]Type) (*COREReloResult, error) {
+	typ := relo.Type
+
+	var localName string
+	if n, ok := typ.(namer); ok {
+		localName = n.name()
+	}
+
+	localSpec, err := coreParseSpec(typ, relo.Accessor, relo.Kind)
+	if err != nil {
+		if localName == "" {
+			localName = "<anon>"
+		}
+		return nil, fmt.Errorf("parsing [%v] %T %s + %s failed: %w", typ, typ, localName, relo.Accessor, err)
+	}
+
+	targetRes := &COREReloResult{}
+	var targetSpec *coreSpec
+	// TYPE_ID_LOCAL relo is special and doesn't need candidate search
+	if relo.Kind == reloTypeIDLocal {
+		targetRes = &COREReloResult{
+			Validate: true,
+			OrigVal:  uint32(localSpec.rootType.ID()),
+			NewVal:   uint32(localSpec.rootType.ID()),
```

Comment:
This "relocation" is simply used to get the Type ID from the local BTF spec.

Reply:
I believe `OrigVal` should already have the correct type ID since the compiler generates the BTF. See the [macro description](https://github.com/libbpf/libbpf/commit/cdb21b05e5c59f4c9fb525a140371d9bcbd59644#diff-92cca15bc801d3458b089b6022edfcaaR120) and [LLVM patch](https://reviews.llvm.org/D74668).

Reply:
Ah, this is a roundabout way of triggering a check during relocation? That would be nice in a comment as well. `TYPE_ID_LOCAL relo is special` doesn't tell me that much.